src/Compilers/CSharp/Portable/Symbols/TypeParameterSymbol.cs

```diff
@@ -369,69 +358,76 @@ internal ImmutableArray<NamedTypeSymbol> AllEffectiveInterfacesWithDefinitionUse
-        /// the initial phase of constraint checking where the constraint types may
-        /// still contain invalid or duplicate types.
         /// </summary>
-        internal abstract void EnsureAllConstraintsAreResolved(bool early);
+        internal abstract void EnsureAllConstraintsAreResolved();
 
         /// <summary>
         /// Helper method to force type parameter constraints to be resolved.
         /// </summary>
-        protected static void EnsureAllConstraintsAreResolved(ImmutableArray<TypeParameterSymbol> typeParameters, bool early)
+        protected static void EnsureAllConstraintsAreResolved(ImmutableArray<TypeParameterSymbol> typeParameters)
         {
             foreach (var typeParameter in typeParameters)
             {
                 // Invoke any method that forces constraints to be resolved.
-                var unused = typeParameter.GetConstraintTypes(ConsList<TypeParameterSymbol>.Empty, early);
+                var unused = typeParameter.GetConstraintTypes(ConsList<TypeParameterSymbol>.Empty);
             }
         }
 
-        internal abstract ImmutableArray<TypeWithAnnotations> GetConstraintTypes(ConsList<TypeParameterSymbol> inProgress, bool early);
+        internal abstract ImmutableArray<TypeWithAnnotations> GetConstraintTypes(ConsList<TypeParameterSymbol> inProgress);
 
         internal abstract ImmutableArray<NamedTypeSymbol> GetInterfaces(ConsList<TypeParameterSymbol> inProgress);
 
         internal abstract NamedTypeSymbol GetEffectiveBaseClass(ConsList<TypeParameterSymbol> inProgress);
 
         internal abstract TypeSymbol GetDeducedBaseType(ConsList<TypeParameterSymbol> inProgress);
 
+        private static bool ConstraintImpliesReferenceType(TypeSymbol constraint)
+        {
+            if (constraint.TypeKind == TypeKind.TypeParameter)
+            {
+                return IsReferenceTypeFromConstraintTypes(((TypeParameterSymbol)constraint).ConstraintTypesNoUseSiteDiagnostics);
+            }
+            else if (!constraint.IsReferenceType)
+            {
+                return false;
+            }
+            else
+            {
+                switch (constraint.TypeKind)
+                {
+                    case TypeKind.Interface:
+                        return false; // can be satisfied by value types
+                    case TypeKind.Error:
+                        return false;
+                }
+
+                switch (constraint.SpecialType)
+                {
+                    case SpecialType.System_Object:
+                    case SpecialType.System_ValueType:
+                    case SpecialType.System_Enum:
+                        return false; // can be satisfied by value types
+                }
+
+                return true;
+            }
+        }
+
         // From typedesc.cpp :
         // > A recursive helper that helps determine whether this variable is constrained as ObjRef.
```

Comment:
Is this comment relevant anymore?

Reply:
> Is this comment relevant anymore?

I think the method is still recursive, indirectly. In any case, I guess this comment is on this method from the beginning of times and refers to relevant parts of native compiler implementation.

---
In reply to: [280542830](https://github.com/dotnet/roslyn/pull/35375#discussion_r280542830) [](ancestors = 280542830)